src/variational/advi.jl

```diff
@@ -71,43 +69,58 @@ function meanfield(rng::AbstractRNG, model::Model)
             ranges[range_idx] = idx .+ r
             range_idx += 1
         end
-        
+
         # append!(ranges, [idx .+ r for r ∈ varinfo.metadata[sym].ranges])
         idx += varinfo.metadata[sym].ranges[end][end]
     end
 
     # initial params
     μ = randn(rng, num_params)
-    σ = softplus.(randn(rng, num_params))
+    σ = StatsFuns.softplus.(randn(rng, num_params))
 
     # construct variational posterior
-    d = TuringDiagMvNormal(μ, σ)
-    bs = inv.(bijector.(tuple(dists...)))
-    b = Stacked(bs, ranges)
+    d = DistributionsAD.TuringDiagMvNormal(μ, σ)
+    bs = inv.(Bijectors.bijector.(tuple(dists...)))
+    b = Bijectors.Stacked(bs, ranges)
 
-    return transformed(d, b)
+    return Bijectors.transformed(d, b)
 end
 
-
 # Overloading stuff from `AdvancedVI` to specialize for Turing
-AdvancedVI.update(d::TuringDiagMvNormal, μ, σ) = TuringDiagMvNormal(μ, σ)
-AdvancedVI.update(td::TransformedDistribution, θ...) = transformed(AdvancedVI.update(td.dist, θ...), td.transform)
-function AdvancedVI.update(td::TransformedDistribution{<:TuringDiagMvNormal}, θ::AbstractArray)
+function AdvancedVI.update(d::DistributionsAD.TuringDiagMvNormal, μ, σ)
+    return DistributionsAD.TuringDiagMvNormal(μ, σ)
+end
+function AdvancedVI.update(td::Bijectors.TransformedDistribution, θ...)
+    return Bijectors.transformed(AdvancedVI.update(td.dist, θ...), td.transform)
+end
+function AdvancedVI.update(
+    td::Bijectors.TransformedDistribution{<:DistributionsAD.TuringDiagMvNormal},
+    θ::AbstractArray,
+)
     μ, ω = θ[1:length(td)], θ[length(td) + 1:end]
-    return AdvancedVI.update(td, μ, softplus.(ω))
+    return AdvancedVI.update(td, μ, StatsFuns.softplus.(ω))
 end
 
-function AdvancedVI.vi(model::Model, alg::ADVI; optimizer = TruncatedADAGrad())
+function AdvancedVI.vi(
+    model::DynamicPPL.Model,
+    alg::AdvancedVI.ADVI;
+    optimizer = AdvancedVI.TruncatedADAGrad(),
+)
```

Comment:
Maybe this one and the one below as well?

Reply:
Hmm, not quite clear to me why we want to move this AVI. In my mind Turing depends on AVI for it's VI-functionality, and so Turing should be the one who overloads accordingly to ensure that the types used by Turing work nicely with AVI (which in this case is `DynamicPPL.Model`).

Reply:
Yeah, I'm less certain about this one. It's just that neither the function nor any of its arguments belong to Turing, which is a bit strange (and type piracy :pirate_flag:)

Reply:
But isn't this more an artifact of Turing now mostly being just a "shell" / "interface" for a lot of upstream packages? Though I agree type-piracy is bad, it seems like we're inevitably going to have to do a bit of it in Turing unless we want to construct a ton of wrapper structs with no real additional functionality. Maybe that's just me being overly pessimistic :shrug: 

Though I agree Requires.jl is an alternative, but it also means that packages that are really upstream for Turing now need to be the ones maintaining compatibility rather than the other way around. I guess that's okay since we're the ones maintaining all the packages, but I feel like the significant use of Requires.jl will come back and bite us at some point. It's already kind of difficult to track down all the upstream packages which require change. In particular it does kind of break the "separation of concerns" as maintainers now need a fairly good knowledge of _all_ the packages involved.

I guess my point is that the type-piracy vs. Requires.jl isn't a clear-cut "always go with Requires.jl"? Maaaybe that's just me though :shrug:

Reply:
No, I'm not convinced either, and in general I'm not a big fan of Requires.jl, it leads to all kind of issues with compatibilities, multiple optional dependencies, package load times, and testing. IMO the Julia package ecosystem still has to come up with a better solution for this general problem.